feat: allow awaiting a page that gets replaced

JaspervanRiet opened the pull request:
## Description

When we navigate to a page and await its result, and that page gets replaced, the following pop is no longer listened to.  This PR implements behaviour where popping from a page that replaced another, will still trigger the original `await navigate`. It does so by creating a new mapping between the new incoming location (the one that replaces) and the completer of the original location when replacing.

## Related Issues

#47 

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is _not_ a breaking change.

DelcoigneYves reviewed (approved):
This works as well! My only concern here is, what if the user is expecting type T to be returned from page B, but since page B is replaced by page C, it returns type U. Is this something that we need to address in the documentation?

Also, since this effectively changes the behavior (for the better, but still), shouldn't we consider this a breaking change?

JaspervanRiet commented:
Thank you, these are good questions. 

> My only concern here is, what if the user is expecting type T to be returned from page B, but since page B is replaced by page C, it returns type U. Is this something that we need to address in the documentation?

Yes, valid concerns. I've added docs, and an exception explaining it. Also added more tests, please re-review.

> Also, since this effectively changes the behavior (for the better, but still), shouldn't we consider this a breaking change?

It does change the behaviour, and I had a think about this, but I'm leaning towards not considering it a breaking change. I think the fact that the completer was never completed makes this a bug, or maybe unsupported, as opposed to behavior that can be built upon.

DelcoigneYves reviewed (approved):
LGTM!
